### PR TITLE
OpenDict

### PR DESCRIPTION
**Pull request by timwoj**
This is based on the work by @jasonlue on https://github.com/zeek/zeek/pull/578 to introduce a new type of dictionary that improves the performance of our existing dictionary, using a new method of hashing.

This PR doesn't currently pass CI due to some timing issues in a few cluster and intel tests, but those are being worked on.

### CODE REVIEW

**Comment by timwoj** [comment]
I just rebased this on top of the current master, which required updating a bunch of test baselines again. Current status:

1. Debug builds still crash a few tests due to calls to `assert()` getting hit that shouldn't be. Notably, the external zeek-testing m57-long test crashes:

```
  % cat .stderr
  1258563581.737779 error in /Users/tim/Desktop/projects/zeek/scripts/policy/protocols/smtp/detect-suspicious-orig.zeek, line 41: Zeek was not configured for GeoIP support (lookup_location(SMTP::ip))
  Assertion failed: (num_iterators == 0 || (cookies && cookies->size() == num_iterators)), function Remove, file ../src/Dict.cc, line 872.
  gunzip: error writing to output: Broken pipe
  cat: stdout: Broken pipe
  /bin/sh: line 1: 69641 Done(1)                 cat $TRACES/2009-M57-day11-18.trace.gz
       69643                       | gunzip
       69644 Abort trap: 6           | zeek -r - /Users/tim/Desktop/projects/zeek/testing/external/zeek-testing/.tmp/tests.m57-long/m57-long.zeek
```

2. Still failing a number of intel and cluster tests because of unstable ordering issues.

**Comment by rsmmr** [comment]
Thanks for updating the branch, @timwoj . @jasonlue can you help address the remaining issues to get it past CI? looks like we're close.

**Comment by timwoj** [comment]
Here's some benchmarking output, just for kicks:

File-read (`-r`) of 5 minutes of datacenter-like data at 500Mbps:
Master:
Time spent: 237.03 seconds
Max memory usage: 2233523 KB

Open-dict:
Time spent: 224.98 seconds
Max memory usage: 2147236 KB

Interface-read (`-i`) of 5 minutes of http/dns data at 500Mbps:
Master:
Maximum memory usage (max_rss): 2870728 KB
Average CPU usage: 74.2%

Open-dict:
Maximum memory usage (max_rss): 2727215 KB
Average CPU usage: 73.1%

**Comment by rsmmr** [comment]
That looks quite promising. I think that means we should still target integration,  however it seems too late now to wrap it up for 3.2.0. I'd rather get this some exercising before turning it into a release. Setting milestone to 4.0, let's aim to merge it early in the cycle.

**Comment by jsiwek** [comment]
I've continued this in `topic/jsiwek/open-dict` in repos: `zeek`, `zeek-docs`, and `zeek-testing-private` and stabilized tests/CI as shown here:

https://cirrus-ci.com/build/6596515990863872

There were changes required to deal with the potential for differing dictionary orderings now, but it was mostly widespread changes to the tests to try to make them more stable and less sensitive to unrelated timing changes (like the new Dict implementation causes indirectly).  A notable bug that's also not related to the new dictionary implementation, but uncovered along the way is this:

https://github.com/zeek/zeek/commit/9fa4769ba956e1071fcf898c251ee845772ddd0b

How to proceed?  Could merge my branch into this PR, start a new PR, or just have someone merge my branch directly if we think it's ready.

**Comment by timwoj** [comment]
Awesome, thanks Jon. I'll cherry-pick this over into the branch.

File: src/Dict.h [hunk 1]
```diff
@@ -2,79 +2,205 @@
...
 // Type for function to be called when deleting elements.
 typedef void (*dict_delete_func)(void*);
 
 namespace zeek {
 
-// Type indicating whether the dictionary should keep track of the order
-// of insertions.
+// Default number of hash buckets in dictionary.  The dictionary will increase the size
+// of the hash table as needed.
+constexpr uint32_t HASH_MASK = 0xFFFFFFFF; //only lower 32 bits.
```
Comment (rsmmr) [inline src/Dict.h:20]: Looks like these constants can move into `detail`?

File: src/Dict.h [hunk 2]
```diff
@@ -2,79 +2,205 @@
...
+
+// These four variables can be used to build different targets with -Dxxx for performance
+// or for debugging purposes.
+
+// When incrementally resizing and remapping, it remaps DICT_REMAP_ENTRIES each step. Use
+// 2 for debug. 16 is best for a release build.
+#ifndef DICT_REMAP_ENTRIES
```
Comment (rsmmr) [inline src/Dict.h:27]: I'd take out the `#ifndef`, very unlikely anybody will ever want to change set any of them; and if so, can always patch the source.

**Pull request merged**